### PR TITLE
JCLOUDS-1005: Backblaze B2 large file operations

### PR DESCRIPTION
**Pull request by andrewgaul**

### CODE REVIEW

**Comment by andrewgaul** [comment]
@nacx @zack-shoylev Please review.

File: b2/src/main/java/org/jclouds/b2/domain/ListUnfinishedLargeFilesResponse.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      public abstract String bucketId();
+      public abstract String contentType();
+      public abstract String fileId();
+      public abstract Map<String, String> fileInfo();
+      public abstract String fileName();
+      public abstract Date uploadTimestamp();
+
+      @SerializedNames({"accountId", "bucketId", "contentType", "fileId", "fileInfo", "fileName", "uploadTimestamp"})
+      public static Entry create(String accountId, String bucketId, String contentType, String fileId, Map<String, String> fileInfo, String fileName, long uploadTimestamp) {
+         return new AutoValue_ListUnfinishedLargeFilesResponse_Entry(accountId, bucketId, contentType, fileId, fileInfo, fileName, new Date(uploadTimestamp));
```
Comment (nacx) [inline b2/src/main/java/org/jclouds/b2/domain/ListUnfinishedLargeFilesResponse.java]: Make the fileInfo map immutable.

File: b2/src/main/java/org/jclouds/b2/domain/MultipartUploadResponse.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+   public abstract String bucketId();
+   public abstract String contentType();
+   public abstract String fileId();
+   public abstract Map<String, String> fileInfo();
+   public abstract String fileName();
+   public abstract Date uploadTimestamp();
+
+   @SerializedNames({"accountId", "bucketId", "contentType", "fileId", "fileInfo", "fileName", "uploadTimestamp"})
+   public static MultipartUploadResponse create(String accountId, String bucketId, String contentType, String fileId, Map<String, String> fileInfo, String fileName, long uploadTimestamp) {
+      return new AutoValue_MultipartUploadResponse(accountId, bucketId, contentType, fileId, fileInfo, fileName, new Date(uploadTimestamp));
```
Comment (nacx) [inline b2/src/main/java/org/jclouds/b2/domain/MultipartUploadResponse.java]: Make the fileInfo map immutable.

File: b2/src/main/java/org/jclouds/b2/features/MultipartApi.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+import org.jclouds.b2.domain.ListUnfinishedLargeFilesResponse;
+import org.jclouds.b2.domain.MultipartUploadResponse;
+import org.jclouds.b2.domain.UploadPartResponse;
+import org.jclouds.b2.filters.RequestAuthorization;
+import org.jclouds.rest.annotations.MapBinder;
+import org.jclouds.rest.annotations.PayloadParam;
+import org.jclouds.rest.annotations.RequestFilters;
+import org.jclouds.rest.binders.BindToJsonPayload;
+
+@BlobScope(CONTAINER)
```
Comment (nacx) [inline b2/src/main/java/org/jclouds/b2/features/MultipartApi.java]: All methods consume json, to the consumes annotation could be moved to class level.

File: b2/src/test/java/org/jclouds/b2/features/MultipartApiMockTest.java [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    *           The name of the resource.
+    * @return The content of the resource
+    */
+   public String stringFromResource(String resourceName) {
+      try {
+         return Strings2.toStringAndClose(getClass().getResourceAsStream(resourceName));
+      } catch (IOException e) {
+         throw Throwables.propagate(e);
+      }
+   }
```
Comment (nacx) [inline b2/src/test/java/org/jclouds/b2/features/MultipartApiMockTest.java]: Create a base class for all mock tests and move the common logic there: api creation, common assertions, etc.

**Comment by nacx** [comment]
Thanks @andrewgaul!

File: b2/src/main/java/org/jclouds/b2/domain/ListUnfinishedLargeFilesResponse.java (hunk 1 above)
Comment (andrewgaul) [inline b2/src/main/java/org/jclouds/b2/domain/ListUnfinishedLargeFilesResponse.java]: Done.

File: b2/src/main/java/org/jclouds/b2/domain/MultipartUploadResponse.java (hunk 2 above)
Comment (andrewgaul) [inline b2/src/main/java/org/jclouds/b2/domain/MultipartUploadResponse.java]: Done.

File: b2/src/main/java/org/jclouds/b2/features/MultipartApi.java (hunk 3 above)
Comment (andrewgaul) [inline b2/src/main/java/org/jclouds/b2/features/MultipartApi.java]: Done.

File: b2/src/test/java/org/jclouds/b2/features/MultipartApiMockTest.java (hunk 4 above)
Comment (andrewgaul) [inline b2/src/test/java/org/jclouds/b2/features/MultipartApiMockTest.java]: Done.

**Comment by andrewgaul** [comment]
Addressed all outstanding review comments.

File: b2/src/main/java/org/jclouds/b2/features/MultipartApi.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+   @POST
+   @MapBinder(UploadPartBinder.class)
+   UploadPartResponse uploadPart(@PayloadParam("response") GetUploadPartResponse response, @HeaderParam("X-Bz-Part-Number") int partNumber, @HeaderParam("X-Bz-Content-Sha1") String sha1, @PayloadParam("payload") Payload payload);
+
+   @Named("b2_list_parts")
+   @POST
+   @Path("/b2api/v1/b2_list_parts")
+   @RequestFilters(RequestAuthorization.class)
+   @MapBinder(BindToJsonPayload.class)
+   @Consumes(APPLICATION_JSON)
```
Comment (nacx) [inline b2/src/main/java/org/jclouds/b2/features/MultipartApi.java]: Remove

File: b2/src/main/java/org/jclouds/b2/features/MultipartApi.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+   @Produces(APPLICATION_JSON)
+   ListPartsResponse listParts(@PayloadParam("fileId") String fileId, @PayloadParam("startPartNumber") @Nullable Integer startPartNumber, @PayloadParam("maxPartCount") @Nullable Integer maxPartCount);
+
+   @Named("b2_list_unfinished_large_files")
+   @POST
+   @Path("/b2api/v1/b2_list_unfinished_large_files")
+   @RequestFilters(RequestAuthorization.class)
+   @MapBinder(BindToJsonPayload.class)
+   @Consumes(APPLICATION_JSON)
```
Comment (nacx) [inline b2/src/main/java/org/jclouds/b2/features/MultipartApi.java]: Remove

**Comment by nacx** [comment]
Two very minors left. Feel free to fix them, squash and merge!

File: b2/src/main/java/org/jclouds/b2/features/MultipartApi.java (hunk 6 above)
Comment (andrewgaul) [inline b2/src/main/java/org/jclouds/b2/features/MultipartApi.java]: Done.

File: b2/src/main/java/org/jclouds/b2/features/MultipartApi.java (hunk 5 above)
Comment (andrewgaul) [inline b2/src/main/java/org/jclouds/b2/features/MultipartApi.java]: Done.  Sorry I missed these!

**Pull request merged**